### ISSUE TITLE
TPU `IsFusibleUnalignedDUS` crash on LLaMA rotate-half RoPE at head_dim=128

### ISSUE THREAD

**Issue by pedronahum** [jax-ml/jax#37029]
### Description

# XLA:TPU `IsFusibleUnalignedDUS` crash on LLaMA rotate-half RoPE at head_dim=128

**Date:** 2026-04-21
**Backend:** TPU v6e (single chip, Google Cloud `v6e-1`)
**Severity:** Hard crash (`CHECK` failure) — not a recoverable error,
the process SIGABRTs.

## TL;DR

A plain `jax.jit(rotate_half_rope)` crashes `libtpu`'s fusion emitter
with

```
Check failed: fusion_util::IsFusibleUnalignedDUS(user, target_)
(fusion_emitter.cc:8881)
```

when the per-head dim is **128** (so the rotate-half split produces
HalfDim=64). The same code **compiles and runs fine at head_dim=256**
(HalfDim=128, MXU-lane-aligned). The crash is independent of batch,
sequence length, head count, and dtype (f32 and bf16 both crash).
Removing the final `jnp.concatenate` (returning the two halves as a
tuple) also avoids the crash — the trigger is specifically the
`split → rotate → concat` pipeline when the half dim isn't MXU-aligned.

This is a critical shape for transformer inference — **D=128 is the
LLaMA-3-8B / LLaMA-2-7B per-head dim**, so any training or inference
script doing RoPE at that shape dies at compile time on this libtpu.

## Environment

```
python  3.11.13
jax     0.10.0
jaxlib  0.10.0
libtpu  0.0.40
TPU     v6e-1 (single-chip Cloud TPU VM)
```

PJRT reports: `PJRT plugin has PJRT API version 0.103. The framework
PJRT API version is 0.96.`

## Minimal repro

```python
# xla_tpu_rope_d128_repro.py
import os
os.environ.setdefault("TPU_LIBRARY_PATH",
                      "/path/to/libtpu/libtpu.so")
os.environ.pop("XLA_FLAGS", None)

import jax
import jax.numpy as jnp

def rope_apply(x, cos, sin):
    """LLaMA rotate-half RoPE. x:[B,S,H,D]; cos,sin:[S,D/2]."""
    x1, x2 = jnp.split(x, 2, axis=-1)
    cos_b = cos[None, :, None, :]
    sin_b = sin[None, :, None, :]
    y1 = x1 * cos_b - x2 * sin_b
    y2 = x1 * sin_b + x2 * cos_b
    return jnp.concatenate([y1, y2], axis=-1)


B, S, H, D = 1, 2048, 32, 128   # <-- D=128 is the LLaMA-3-8B head dim
HD = D // 2

x   = jnp.zeros((B, S, H, D),  dtype=jnp.bfloat16)
cos = jnp.zeros((S, HD),       dtype=jnp.bfloat16)
sin = jnp.zeros((S, HD),       dtype=jnp.bfloat16)

fn = jax.jit(rope_apply)
print(f"compiling at B={B} S={S} H={H} D={D} HD={HD} dtype=bf16...")
out = fn(x, cos, sin)
out.block_until_ready()
print(f"ok, shape={out.shape}, dtype={out.dtype}")
```

Run:

```
python3.11 xla_tpu_rope_d128_repro.py
```

## Exact output observed

```
compiling at B=1 S=2048 H=32 D=128 HD=64 dtype=bf16...
F0421 12:42:19.646137 2171764 fusion_emitter.cc:8881] Check failed: fusion_util::IsFusibleUnalignedDUS(user, target_)
*** Check failure stack trace: ***
    @     0x...  absl::log_internal::LogMessage::Fail()
    @     0x...  xla::jellyfish::FusionEmitter::EstimateFusionCost()
    @     0x...  xla::jellyfish::FusionEmitter::EstimateTotalCycles()
    @     0x...  xla::jellyfish::FusionEmitter::IterateThroughWindowConfigs()
    @     0x...  xla::jellyfish::FusionEmitter::ComputeWindowConfigInternal()
    @     0x...  xla::jellyfish::FusionEmitter::Emit()
    @     0x...  xla::jellyfish::LoweringEmitter::PrepareOutput()::$_3::operator()()
*** SIGABRT received ...
```

The process exits with signal 6 (ABRT). No Python exception; the
crash is inside libtpu before control returns to JAX.

## What DOES and DOES NOT trigger the crash

Held everything else fixed and swept one variable at a time:

| Variable  | Value         | Status  |
|-----------|---------------|---------|
| D         | 128 (HD=64)   | **crash** |
| D         | 256 (HD=128)  | ok      |
| dtype     | f32           | crash (at D=128) |
| dtype     | bf16          | crash (at D=128) |
| S         | 16            | crash (at D=128) |
| S         | 2048          | crash (at D=128) |
| batch B   | 1             | crash (at D=128) |
| heads H   | 32            | crash (at D=128) |
| pipeline  | split + rotate + concat  | crash (at D=128) |
| pipeline  | split + rotate (no concat) | ok (returns tuple) |

**So the specific trigger is:** `split → rotate → concat` where the
result of split has a trailing dim that is not MXU-lane-aligned (64,
which is a natural byproduct of D=128 with half-split RoPE).

## What the generated HLO looks like

At `B=1, S=2048, H=32, D=128, dtype=bf16` the StableHLO from
`jax.jit(rope_apply).lower(...)` is (abbreviated):

```
%arg0: tensor<1x2048x32x128xbf16>
%arg1, %arg2: tensor<2048x64xbf16>   -- cos, sin

%x1  = stablehlo.slice %arg0 [0:1, 0:2048, 0:32, 0:64]   -> tensor<1x2048x32x64xbf16>
%x2  = stablehlo.slice %arg0 [0:1, 0:2048, 0:32, 64:128] -> tensor<1x2048x32x64xbf16>
%cos_b = stablehlo.broadcast_in_dim %arg1, dims = [1, 3] -> tensor<1x2048x32x64xbf16>
%sin_b = stablehlo.broadcast_in_dim %arg2, dims = [1, 3] -> tensor<1x2048x32x64xbf16>
%y1 = %x1 * %cos_b - %x2 * %sin_b
%y2 = %x1 * %sin_b + %x2 * %cos_b
%out = stablehlo.concatenate %y1, %y2, dim = 3  -> tensor<1x2048x32x128xbf16>
```

XLA presumably rewrites the concat-of-two-halves into a pair of
`dynamic_update_slice`s for efficiency, and then the fusion emitter
trips when it tries to fuse users against a DUS whose destination
slice starts at an offset not aligned to the MXU lane count (128).

## Workarounds

1. **Pad D to an MXU-aligned value.** We moved our bench to D=256
   (HalfDim=128 matches the v6e MXU lane width) and the same code
   compiles and runs at ~0.15 ms. Not viable for production
   LLaMA-3-8B / -2-7B / -3-70B inference — all three use
   head_dim=128.
2. **Return the halves as a tuple** (don't concatenate inside the
   jit). The caller then has to do the concat in a separate jit or
   eagerly, losing the fusion.
3. **Use an interleaved / complex-multiply RoPE variant** rather
   than rotate-half. Not tested here, but wouldn't hit the same
   concat-of-halves pattern. Larger behavioral change for anyone
   whose checkpoint was trained with rotate-half freqs.

## Environment

```
python  3.11.13
jax     0.10.0
jaxlib  0.10.0
libtpu  0.0.40
TPU     v6e-1 (single-chip Cloud TPU VM)
```
